indico/modules/events/registration/client/js/form/fields/DateInput.jsx

```diff
@@ -8,69 +8,167 @@
+            required={required}
+            showSecond={false}
+            value={toMoment(timeValue, 'HH:mm:ss', true)}
+            focusOnOpen
+            onChange={handleTimeChange}
+            use12Hours={timeFormat === '12h'}
+            allowEmpty={false}
+            placeholder={timeFormat === '12h' ? '--:-- am/pm' : '--:--'}
+            getPopupContainer={node => node}
+          />
+        </Form.Field>
+      )}
+    </Form.Group>
+  );
+}
+
+DateInputComponent.propTypes = {
+  value: PropTypes.string.isRequired,
+  onChange: PropTypes.func.isRequired,
+  disabled: PropTypes.bool.isRequired,
+  required: PropTypes.bool.isRequired,
+  dateFormat: PropTypes.oneOf([
+    'DD/MM/YYYY',
+    'DD.MM.YYYY',
+    'MM/DD/YYYY',
+    'MM.DD.YYYY',
+    'YYYY/MM/DD',
+    'YYYY.MM.DD',
+    'MM/YYYY',
+    'MM.YYYY',
+    'YYYY',
+  ]).isRequired,
+  timeFormat: PropTypes.oneOf(['12h', '24h']),
+};
+
+DateInputComponent.defaultProps = {
+  timeFormat: null,
+};
+
+export default function DateInput({htmlName, disabled, isRequired, dateFormat, timeFormat}) {
   const friendlyDateFormat = dateFormat.replace(
     /%([HMdmY])/g,
     (match, c) => ({H: 'HH', M: 'mm', d: 'DD', m: 'MM', Y: 'YYYY'}[c])
   );
+  const validateDateTime = dateTime => {
+    if (dateTime && !dateTime.match(/\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:00(.*)/)) {
+      return Translate.string('The provided date is invalid.');
+    }
+  };
 
   if (dateFormat.includes('%d')) {
     return (
-      <Form.Group styleName="date-field">
-        <Form.Field>
-          <SingleDatePicker
-            id={`regform-datepicker-${id}`}
-            disabled={disabled}
-            date={date}
-            onDateChange={setDate}
-            placeholder={friendlyDateFormat}
-            displayFormat={friendlyDateFormat}
-            isOutsideRange={() => false}
-            verticalSpacing={10}
-            enableOutsideDays
-          />
-        </Form.Field>
-        {timeFormat && (
-          <Form.Field>
-            <TimePicker
-              id={`regform-timepicker-${id}`}
-              disabled={disabled}
-              showSecond={false}
-              value={time}
-              focusOnOpen
-              onChange={setTime}
-              use12Hours={timeFormat === '12h'}
-              allowEmpty={false}
-              placeholder={timeFormat === '12h' ? '--:-- am/pm' : '--:--'}
-              getPopupContainer={node => node}
-            />
-          </Form.Field>
-        )}
-      </Form.Group>
+      <FinalField
+        name={htmlName}
+        component={DateInputComponent}
+        required={isRequired}
+        disabled={disabled}
+        dateFormat={friendlyDateFormat}
+        timeFormat={timeFormat}
+        validate={validateDateTime}
+      />
     );
   } else {
+    const parseDate = date => {
+      if (!date) {
+        return date;
+      }
+      const numbers = date.replace(/[^\d]/g, '');
+      if (numbers.length >= 4 && dateFormat === '%Y') {
```

Comment:
Shouldn't >4 be considered invalid?

Reply:
Maybe this whole validation logic could be simplified a bit by trying to parse the user input with the given time format, and then using momentjs to format it to the iso format? The only cases that need special handling would by those with just year/month or just year.

Reply:
I didn't have much luck when I tried to do that with momentjs... It seems to accept dates that are incomplete and throws everything off...

What this logic does is it parses the input as the user is typing, disallowing most invalid inputs from the get go.
The `>= 4` is basically limiting the field value length - I could replace it with a `=== 4` if I set the `maxlength` on the input, but this way seemed easier

Reply:
What i mean is that currently something like `2022x` would probably be accepted (since the garbage at the end would just be ignored).

The fact that momentjs accepts nonsense even when providing a specific format is WTF. Not sure what's the best solution there tbh... maybe a mapping between the format and a regexp to parse the string?

Reply:
something like this:

```js
const regexps = {
  'DD.MM.YYYY': /^(?<day>\d{2})\.(?<month>\d{2})\.(?<year>\d{4})$/,
  ...
};
```

And then you can .exec() it and get the named groups in its `.groups` attribute - that way you have standard data that should be easy to validate or pass to momentjs (mainly to make sure the data isn't some nonsense like 99.99.9999)